modules/apps/journal/journal-service/src/main/java/com/liferay/journal/internal/search/JournalArticleIndexer.java

```diff
@@ -151,17 +152,18 @@ public SearchContext getSearchContext() {
 			});
 	}
 
-	@Reference(
-		cardinality = ReferenceCardinality.MULTIPLE,
-		policy = ReferencePolicy.DYNAMIC,
-		policyOption = ReferencePolicyOption.GREEDY,
-		service = ModelDocumentContributor.class,
-		target = "(indexer.class.name=com.liferay.journal.model.JournalArticle)"
-	)
-	protected void addModelDocumentContributor(
-		ModelDocumentContributor<JournalArticle> modelDocumentContributor) {
+	@Activate
+	protected void activate(BundleContext bundleContext) {
+		_serviceTrackerList = ServiceTrackerListFactory.open(
+			bundleContext,
+			(Class<ModelDocumentContributor<JournalArticle>>)
```

Comment:
@JanisZhang How many `ModelDocumentContributor` trackers with different indexer.class.name filters do we have in total?

Reply:
Hi @dantewang,

Three classes are tracking `ModelDocumentContributor<T extends BaseModel>`:  KBArticleIndexer, JournalArticleIndexer, and WikiPageIndexer. 

All of them extend the class `BaseIndex` , but in a different module. I tried to combine them into a registry but I can not find a good way...

`JournalArticleIndexer`: this 
`KBArticleIndexer`:      https://github.com/liferay-core-infra/liferay-portal/pull/4000
`WikiPageIndexer`:      https://github.com/liferay-core-infra/liferay-portal/pull/3996